llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -592,8 +662,223 @@ static Error applyAddends(MCCASReader &Reader,
   return Error::success();
 }
 
-Expected<uint64_t> DebugInfoSectionRef::materialize(MCCASReader &Reader,
-                                                    raw_ostream *Stream) const {
+static Expected<uint64_t> getFormSize(dwarf::Form FormVal, dwarf::FormParams FP,
+                                      ArrayRef<char> CUData, uint64_t CUOffset,
+                                      bool IsLittleEndian,
+                                      uint8_t AddressSize) {
+  uint64_t FormSize = 0;
+  bool Indirect = false;
+  Error Err = Error::success();
+  do {
+    Indirect = false;
+    switch (FormVal) {
+    case dwarf::DW_FORM_addr:
+    case dwarf::DW_FORM_ref_addr: {
+      FormSize += (FormVal == dwarf::DW_FORM_addr) ? FP.AddrSize
+                                                   : FP.getRefAddrByteSize();
+      break;
+    }
+    case dwarf::DW_FORM_exprloc:
+    case dwarf::DW_FORM_block: {
+      DWARFDataExtractor DWARFExtractor(toStringRef(CUData), IsLittleEndian,
+                                        AddressSize);
+      uint64_t PrevOffset = CUOffset;
+      CUOffset += DWARFExtractor.getULEB128(&CUOffset);
+      FormSize += CUOffset - PrevOffset;
```

Comment:
What happens if the ULEB (or any form really) extends past the end of CUData? Do we catch this?
This will happen when there are bugs that e.g, introduce an off-by-one error, and the earlier we catch them the easier to debug

Reply:
Ideally the ULEB would not be encoded so that we go past the end, but yes we should check

Reply:
I noticed that every get* function in DWARFExtractor can take an Error as a second argument, so that should be enough to prevent any out of bounds reads.

Reply:
I also made sure to return an error if the FormSize that is returned is too big for the next copy into the buffer that contains the contents for the debug info section